File: samples/Java/animate-a-pendulum.java
import java.awt.*;
import javax.swing.*;

public class Pendulum extends JPanel implements Runnable {

    private double angle = Math.PI / 2;
    private int length;

    public Pendulum(int length) {
        this.length = length;
        setDoubleBuffered(true);
    }

    @Override
    public void paint(Graphics g) {
        g.setColor(Color.WHITE);
        g.fillRect(0, 0, getWidth(), getHeight());
        g.setColor(Color.BLACK);
        int anchorX = getWidth() / 2, anchorY = getHeight() / 4;
        int ballX = anchorX + (int) (Math.sin(angle) * length);
        int ballY = anchorY + (int) (Math.cos(angle) * length);
        g.drawLine(anchorX, anchorY, ballX, ballY);
        g.fillOval(anchorX - 3, anchorY - 4, 7, 7);
        g.fillOval(ballX - 7, ballY - 7, 14, 14);
    }

    public void run() {
        double angleAccel, angleVelocity = 0, dt = 0.1;
        while (true) {
            angleAccel = -9.81 / length * Math.sin(angle);
            angleVelocity += angleAccel * dt;
            angle += angleVelocity * dt;
            repaint();
            try { Thread.sleep(15); } catch (InterruptedException ex) {}
        }
    }

    @Override
    public Dimension getPreferredSize() {
        return new Dimension(2 * length + 50, length / 2 * 3);
    }

    public static void main(String[] args) {
        JFrame f = new JFrame("Pendulum");
        Pendulum p = new Pendulum(200);
        f.add(p);
        f.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        f.pack();
        f.setVisible(true);
        new Thread(p).start();
    }
}

File: samples/Scala/animate-a-pendulum.scala
import scala.swing._
import scala.swing.Swing._
import scala.actors._
import scala.actors.Actor._

import java.awt.{Color, Graphics}

object Pendulum extends SimpleSwingApplication {

  val length    = 100

  val prefSizeX = 2*length+50
  val prefSizeY = length/2*3

  lazy val ui = new Panel {
    import math._
    background = Color.white
    preferredSize = (prefSizeX, prefSizeY)
    peer.setDoubleBuffered(true)

    var angle: Double = Pi/2;

    def pendular = new Actor {
      var angleAccel, angleVelocity = 0.0;
      var dt = 0.1

      def act() {
        while (true) {
          angleAccel = -9.81 / length * sin(angle)
          angleVelocity += angleAccel * dt
          angle += angleVelocity * dt
          repaint()
          Thread.sleep(15)
        }
      }
    }

    override def paintComponent(g: Graphics2D) = {
      super.paintComponent(g)

      g.setColor(Color.white);
      g.fillRect(0, 0, size.width, size.height);
      val anchorX = size.width / 2
      val anchorY = size.height / 4
      val ballX = anchorX + (sin(angle) * length).toInt
      val ballY = anchorY + (cos(angle) * length).toInt
      g.setColor(Color.lightGray)
      g.drawLine(anchorX-2*length, anchorY, anchorX+2*length, anchorY)
      g.setColor(Color.black)
      g.drawLine(anchorX, anchorY, ballX, ballY)
      g.fillOval(anchorX - 3, anchorY - 4, 7, 7)
      g.drawOval(ballX - 7, ballY - 7, 14, 14)
      g.setColor(Color.yellow)
      g.fillOval(ballX - 7, ballY - 7, 14, 14)
    }
  }

  def top = new MainFrame {
    title = "Rosetta Code >>> Task: Animate a pendulum | Language: Scala"
    contents = ui
    ui.pendular.start
  }

}
